mypy/checker.py

```diff
@@ -1116,6 +1117,10 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                                                       infer_lvalue_type)
         else:
             lvalue_type, index_lvalue, inferred = self.check_lvalue(lvalue)
+
+            if isinstance(lvalue, NameExpr):
+                self.check_compatibility_all_supers(lvalue, lvalue_type, rvalue)
```

Comment:
FWIW it's perhaps easier to understand if you use a different type on the last line. E.g. if you use 0.0 there the errors changes to
```
__tmp__.py:5: error: Incompatible types in assignment (expression has type "str", variable has type "int")
__tmp__.py:5: error: Incompatible types in assignment (expression has type "float", variable has type "str")
```
This also reminds me of a general issue I have with this PR: it would be nice to show the file+line of the conflicting variable definition in the base class. (I vaguely recall asking for that before?)

And that then reminds me of another thing: since your loop over the MRO breaks at the first conflict, I think it would be better to go in forward direction (i.e. drop the `reversed()` call), so that the *nearest* conflict is reported. That seems important in a case like this:
```python
class A:
    x = 0
class B(A):
    x = ''  # type: ignore
class C(B):
    x = 0.0
```
Your code currently reports a conflict between `float` and `int` for class `C`, while (given the `# type: ignore` in class `B`) the more relevant conflict is between `float` and `str`. In fact, I believe this code should not have an error at all:
```python
class A:
    x = 0
class B(A):
    x = ''  # type: ignore
class C(B):
    x = ''  # should be okay
```